### ISSUE TITLE
CasADi: Broaden definition of alias

### ISSUE THREAD

**Issue by PeterRabbit95** [pymoca/pymoca#217]
It would be nice to have an option to simplify equations like the following:

``` python
[MX(((-gridResistance_n_v)-(0.8*circuitToPowerGrid_n_i))),
 MX((gridResistance_LossPower+(gridResistance_n_v*circuitToPowerGrid_n_i))),
 MX(((-circuitToPowerBattery_n_v)-(0.086*circuitToPowerBattery_n_i))),
 MX((battery_batteryResistance_LossPower+(circuitToPowerBattery_n_v*circuitToPowerBattery_n_i))),
 MX((der(battery_xSOE)-(3.58474e-06*circuitToPowerBattery_n_i))),
 MX((battery_v-(765-circuitToPowerBattery_n_v))),
 MX(((controlPowerConverter+controlPowerBattery)-(powerRequired-brakingResistor_elPowerInput_powerFlow))),
 MX(((-(circuitToPowerGrid_v*circuitToPowerGrid_n_i))-powerDCDC_elPowerInput_powerFlow)),
 MX((circuitToPowerGrid_v-(680-gridResistance_n_v))),
 MX(((-(circuitToPowerBattery_v*circuitToPowerBattery_n_i))-controlPowerBattery)),
 MX((circuitToPowerBattery_v-(765-circuitToPowerBattery_n_v))),
 MX(@1=0.92, @2=183, (fmin(((@1*powerDCDC_elPowerInput_powerFlow)-@2),((powerDCDC_elPowerInput_powerFlow/@1)-@2))-controlPowerConverter)),
 MX((powerDCDC_powerLosses-(powerDCDC_elPowerInput_powerFlow-controlPowerConverter)))]

```

As you can see most of them are in the form:

`A*x +/- B*y = 0`

It could be very nice to be able to alias `x` to `y` by symbolically solving the relevant equation. 

A possible way to do this would be to extend the capabilities of detect_aliases (even though I am not sure if it is appropriate). I wrote implemented something in this direction in [this commit](https://github.com/PeterRabbit95/pymoca/tree/broad_alias_definition). There is a test with a little example of the functionality.

It is actually a very big change in the logic behind detect_aliases and as for now it works only with simple examples (also it is not backward compatible with the previous logic. I did not even try to run all the tests). 

Before continuing I wanted to share this idea with you @jackvreeken  Do you think it is feasible? Do you think that changing the alias logic can be a good way to achieve this result?

**Comment by jackvreeken** [comment]
I'm not entirely sure I follow. Do you want to store a factor or something in the alias relation between two variables as well? 

```Modelica
0.8 * x - y = 0
```

So something like that would make `x` an alias of `y` with a factor of 0.8. And then `x` would be substituted with `1 / 0.8 * y` everywhere. Am I understanding your intention correctly like that?

**Comment by PeterRabbit95** [comment]
Correct! That's exactly the basic idea. Not only to alias variables with a different sign(+/-) but also with a constant factor.

The problem I see is that the connected variables will likely not represent the same quantity (and unit) anymore. I am not sure if this is something we should avoid in the alias logic and simply do it somewhere else.

In the branch I quoted before I actually went a bit further, changing a lot in the alias logic. As I said it is far from complete but aliases are based on the following class:

```python
class Alias:
    def __init__(self, alias, canonical, direct, inverse):
        self.alias = alias
        self.canonical = canonical
        self.direct = direct
        self.inverse = inverse
        self.direct_call = ca.Function('Direct', [self.canonical], [direct])
        self.inverse_call = ca.Function('Inverse', [self.alias], [inverse])
```

In this way, we could keep track of every possible relation between the connected variables as long as it is explicitable in both directions. For example:

```python
x=ca.MX.sym('x')
y=ca.MX.sym('y')

eq = ca.sin(x) + 9*y

## Here we need a logic to actually compute direct and inverse automatically. I implemented only a few scenarios

Alias(alias = x, canonical=y, direct=ca.asin(-9*y), inverse=-ca.sin(x)/9)
```

**Comment by jackvreeken** [comment]
Well, I refactored the `AliasRelation` class at a certain point to squeeze out as much performance as I could. This is mostly related to quickly being able to look up the alias of any variable (or its negative). I sometimes have to do this millions of times. Leaving aside backwards compatibility of the API (we are not 1.0 yet anyway), I would really like to avoid these calls to become slower for some reason.

It would be pretty thgouh, I think, to have a single alias_relation that does everything. That avoids breaking the changes of #216 . It might make sense to make the detection of "alias with a non-unity factor"-type aliases optional though.

**Comment by PeterRabbit95** [comment]
Hi, sorry I did not reply to this for a while. I looked into `eliminable_variable_expression` and I believe it is a good candidate to achieve what I was talking about. 

I think the big strength of `detect_alias` is being able to keep track of the simplified variables and convert the attributes. On the other hand, the logic behind `eliminable_variable_expression` is more complete but it does not convert the attributes and, as for now, it works only by specifying the names to be eliminated.

Furthermore, something I noticed is that when we eliminate a variable with `replace_parameter_values` or `eliminable_variable_expression` (but this holds true also for other logics) we simply delete it. Instead, when we use `detect_alias` we try to keep track of the different names. Is there a reason for that? Would it make sense to simply delete the direct aliases with `detect_alias` and use a more general logic to keep track of all the simplified variables such as what I implemented in #220?

In conclusion, I think it would be cool to have a common set of simplification methods and call them according to which option is active:
- `detect_alias` only acts on direct aliases
- `eliminable_variable_expression` simplify as much as possible the variables that match the regex logic
- `maximum_simplification` simplify as much as possible every variable

All of these logics should be able to:
- Track the eliminated variables- Track the eliminated variables

- Convert the attributes
